docs: reorganize and separate out HACKING.md for technical instructions

pluiedev opened the pull request:
Also did some copy-editing and removed some things that haven't been present in the codebase for a while (I thought we nuked conformance stuff in 1.1 already...?)

pluiedev committed:
docs: separate out HACKING.md from README.md

Also did some copy-editing and removed some things that haven't been
present in the codebase for a while (I thought we nuked conformance
stuff in 1.1 already...?)

jcollie reviewed (commented):
Maybe instead of a new file, this should be merged into `CONTRRIBUTING.md`?

pluiedev commented:
There's a lot of overlap for sure... Personally I think CONTRIBUTING.md is already *way* too large and I think we can move the technical parts of it over to HACKING.md, and then keep CONTRIBUTING.md as the document that more describes the *process* than the exact technical details.

pluiedev committed:
docs: divide content more evenly between CONTRIBUTING and HACKING

CONTRIBUTING should now solely be about the contribution *process*
while HACKING goes into the technical details

mitchellh commented:
> There's a lot of overlap for sure... Personally I think CONTRIBUTING.md is already _way_ too large and I think we can move the technical parts of it over to HACKING.md, and then keep CONTRIBUTING.md as the document that more describes the _process_ than the exact technical details.

Haven't reviewed yet, but I agree with this assessment. 

CONTRIBUTING should be as small as possible to hopefully encourage as many contributors as possible to read it and follow the rules. It should be focused primarily on guidelines to contribution and not really _how to dev_ itself.

pluiedev committed:
docs: better integrate CONTRIBUTING into the README
